include/ginkgo/core/preconditioner/gauss_seidel.hpp

```diff
@@ -0,0 +1,126 @@
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_PRECONDITIONER_GAUSS_SEIDEL_HPP_
+#define GKO_PUBLIC_CORE_PRECONDITIONER_GAUSS_SEIDEL_HPP_
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/solver/triangular.hpp>
+
+
+namespace gko {
+
+namespace preconditioner {
+template <typename ValueType = default_precision, typename IndexType = int32>
+class GaussSeidel : public EnableLinOp<GaussSeidel<ValueType, IndexType>>,
+                    public Transposable {
+    friend class EnableLinOp<GaussSeidel>;
+    friend class EnablePolymorphicObject<GaussSeidel, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using Csr = matrix::Csr<ValueType, IndexType>;
+    using Dense = matrix::Dense<ValueType>;
+    /*   using LTrs = solver::LowerTrs<value_type, index_type>;
+      using UTrs = solver::UpperTrs<value_type, index_type>; */
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    GaussSeidel& operator=(const GaussSeidel& other);
+
+    GaussSeidel& operator=(GaussSeidel&& other);
+
+    GaussSeidel(const GaussSeidel& other);
+
+    GaussSeidel(GaussSeidel&& other);
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        bool GKO_FACTORY_PARAMETER_SCALAR(skip_sorting, true);
+
+        bool GKO_FACTORY_PARAMETER_SCALAR(symmetric_preconditioner, false);
+
+        double GKO_FACTORY_PARAMETER_SCALAR(relaxation_factor, 1.0);
```

Comment:
```suggestion
        value_type GKO_FACTORY_PARAMETER_SCALAR(relaxation_factor, 1.0);
```
But I think it makes more sense to have two classes, `GaussSeidel` and `Sor`, and move this parameter to `Sor`. Internally they may still use the same kernels